### PR TITLE
 fix(#25): reconcile cluster scope resources

### PR DESCRIPTION
**Pull request by dipak-pawar**
Fixes: #25

### CODE REVIEW

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go [hunk 1]
```diff
@@ -94,8 +96,16 @@ func (r *ReconcileToolChainEnabler) Reconcile(request reconcile.Request) (reconc
 
 	// Fetch the ToolChainEnabler instance
 	instance := &codereadyv1alpha1.ToolChainEnabler{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+	namespacedName := request.NamespacedName
+
+	// overwrite for cluster scoped resources like OAuthClient, ClusterRoleBinding as you can't get namespace from it's event
+	if request.Namespace == "" {
+		log.Info("Couldn't find namespace in the request, getting it from env variable `WATCH_NAMESPACE`")
+		namespacedName = types.NamespacedName{Namespace: os.Getenv("WATCH_NAMESPACE"), Name: request.Name}
```
Comment (alexeykazakov) [inline pkg/controller/toolchainenabler/toolchainenabler_controller.go]: Maybe worth checking if WATCH_NAMESPACE env var is set and log an error if it's not?

**Comment by dipak-pawar** [comment]
> I wonder if we need to watch OAuthClient and ClusterRole at all

We are watching it to cover the case where someone with required permission delete `OAuthClient` or `ClusterRole` accidently, in that case even if we have required cluster info in cluster service, it won't work.

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go [hunk 2]
```diff
@@ -94,8 +96,20 @@ func (r *ReconcileToolChainEnabler) Reconcile(request reconcile.Request) (reconc
...
 	instance := &codereadyv1alpha1.ToolChainEnabler{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+	namespacedName := request.NamespacedName
+
+	// overwrite for cluster scoped resources like OAuthClient, ClusterRoleBinding as you can't get namespace from it's event
+	if request.Namespace == "" {
+		log.Info("Couldn't find namespace in the request, getting it from env variable `WATCH_NAMESPACE`")
+		ns := os.Getenv("WATCH_NAMESPACE")
+		if ns == "" {
+			log.Error(errs.New("env variable `WATCH_NAMESPACE` is not set"), "can't reconcile request coming from cluster scoped resources event")
```
Comment (MatousJobanek) [inline pkg/controller/toolchainenabler/toolchainenabler_controller.go]: is it fine that it is not returned here?

Comment (dipak-pawar) [reply pkg/controller/toolchainenabler/toolchainenabler_controller.go]: yes, as it won't help if you return an error, as it'll again reconcile due to error, so we are just logging the error, so user will know about it.

Comment (MatousJobanek) [reply pkg/controller/toolchainenabler/toolchainenabler_controller.go]: I didn't mean returning an error, just an empty result `return reconcile.Result{}, nil`

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go [hunk 3]
```diff
@@ -94,8 +96,20 @@ func (r *ReconcileToolChainEnabler) Reconcile(request reconcile.Request) (reconc
...
 	// Fetch the ToolChainEnabler instance
 	instance := &codereadyv1alpha1.ToolChainEnabler{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, instance); err != nil {
+	namespacedName := request.NamespacedName
+
+	// overwrite for cluster scoped resources like OAuthClient, ClusterRoleBinding as you can't get namespace from it's event
+	if request.Namespace == "" {
+		log.Info("Couldn't find namespace in the request, getting it from env variable `WATCH_NAMESPACE`")
+		ns := os.Getenv("WATCH_NAMESPACE")
+		if ns == "" {
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/toolchainenabler/toolchainenabler_controller.go]: yes

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go (hunk 2 above)
Comment (dipak-pawar) [reply pkg/controller/toolchainenabler/toolchainenabler_controller.go]: this is a good point. Let me do that.

File: test/e2e/toolchain_enabler_test.go [hunk 4]
```diff
@@ -36,17 +27,26 @@ func TestTooChainEnabler(t *testing.T) {
...
+	err = ctx.InitializeClusterResources(&framework.CleanupOptions{TestContext: ctx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	require.NoError(t, err, "failed to initialize cluster resources: %v", err)
+	t.Log("Initialized cluster resources")
 
-func verifyResources(t *testing.T, f *framework.Framework, ctx *framework.TestCtx) error {
 	namespace, err := ctx.GetNamespace()
-	if err != nil {
-		return fmt.Errorf("could not get namespace: %v", err)
-	}
+	require.NoError(t, err, "failed to get namespace where operator needs to run: %v", err)
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply test/e2e/toolchain_enabler_test.go]: Let me remove error from the message

File: test/e2e/toolchain_enabler_test.go [hunk 5]
```diff
@@ -36,17 +27,26 @@ func TestTooChainEnabler(t *testing.T) {
 	err := framework.AddToFrameworkScheme(apis.AddToScheme, toolChainEnablerList)
 	require.NoError(t, err, "failed to add custom resource scheme to framework: %v", err)
 
-	// run subtests
-	t.Run("Toolchain", func(t *testing.T) {
-		t.Run("Enable", EnableCodeReadyToolChain)
-	})
-}
+	os.Setenv("TEST_NAMESPACE", "toolchain-e2e-test")
+	defer os.Unsetenv("TEST_NAMESPACE")
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply test/e2e/toolchain_enabler_test.go]: thanks for pointing to the code

File: test/e2e/toolchain_enabler_test.go [hunk 6]
```diff
@@ -104,7 +104,7 @@ func TestTooChainEnabler(t *testing.T) {
 
 		// when
 		err = operatorClient.Delete(context.Background(), sa)
-		require.NoError(t, err, "failed to delete service account %s/%s: %s", namespace, toolchainenabler.SAName, err)
+		require.NoError(t, err, "failed to delete service account %s/%s", namespace, toolchainenabler.SAName)
```
Comment (MatousJobanek) [inline test/e2e/toolchain_enabler_test.go]: you maybe wanted to use `require.NoErrorf` same here https://github.com/fabric8-services/toolchain-operator/pull/27/commits/07f0dabeb024fb2de80aaa09f975cce012f1ad38#diff-be3f50b9568b9dffce1c0c348a2fb70bR93

Comment (dipak-pawar) [reply test/e2e/toolchain_enabler_test.go]: I have removed error from the message

Comment (MatousJobanek) [reply test/e2e/toolchain_enabler_test.go]: but you still need to format the message: `"failed to delete service account %s/%s"`

**Comment by MatousJobanek** [comment]
Maybe last note - it would be good to have the `WATCH_NAMESPACE` set as a constant `WatchNamespaceEnvVar` - this is what the other operators do

**Comment by MatousJobanek** [comment]
btw @dipak-pawar - have you tried using `k8sutil.GetWatchNamespace()` function?

File: test/e2e/toolchain_enabler_test.go (hunk 6 above)
Comment (dipak-pawar) [reply test/e2e/toolchain_enabler_test.go]: it's not needed, as internally, it will construct message using `fmt.Sprintf()`

**Comment by dipak-pawar** [comment]
> have you tried using `k8sutil.GetWatchNamespace()` function?

done in a03a656

**Comment by alexeykazakov** [comment]
> > I wonder if we need to watch OAuthClient and ClusterRole at all
> 
> We are watching it to cover the case where someone with required permission delete `OAuthClient` or `ClusterRole` accidently, in that case even if we have required cluster info in cluster service, it won't work.

OK. Makes sense.

**Pull request merged**